### ISSUE TITLE
Allow >1 in-flight authentication sessions

### ISSUE THREAD

**Issue by phavekes** [OpenConext/OpenConext-engineblock#1532]
[This issue is imported from pivotal - Originaly created at Feb 20, 2019 by Thijs Kinkhorst](https://www.pivotaltracker.com/story/show/164121029)

This problem is shown as follows:
* Open a URL that uses OpenConext auth in the first tab (e.g.: Profile). You are redirected to a WAYF. Do not do anything there yet.
* Open a URL that uses OpenConext auth in the second tab (e..g: Teams or Manage). You are again redirected to a (different) WAYF.
* Continue the process (select an idp, log in) in one of the tabs. Be logged in.
* Continue the process in the other tab (select an (can be the same) idp, log in). This will always fail with a hard error.

This story happens e.g. when you\'re browsing and opening two or more links in tabs, links that are protected with a SAML login.

The cause is that Engineblock only keeps track of a single "in flight" authentication session. You need to complete a full sp -> eb -> idp -> eb -> sp cycle before you can do another one.

Obviously this is not user friendly. Make it so that the user can start multiple authentication requests and that all can be completed in any order.

Every authnrequest will have a request ID which is passed to the idp and returned from the idp (InResponseTo). This may be a useful tool to keep them apart.

**Comment by phavekes** [comment]
While working on the Behat tests for this feature I ran into a limitation of the Behat session we are using. We utilize Goutte sessions to run the Behat steps. Goutte, in turn, uses Guzzle/curl to translate the steps into meaningfull http requests. This session type limits us to the capabilities of Curl. So JS interaction or running multiple browser windows in one session is not supported. The latter we realy need in order to simulate the > 1 in-flight authentication scenarios.

I\'m going to experiment with adding an additional Behat session next to the default Goutte session, which utilizes Google Chrome in headless mode to run tests against. This does require some additional dependencies:
 - Selenium 2 :sad_face:
 - ChromeDriver
Both should run fine on Travis

As an alternative we could use BrowserStack (BS) to run the tests against. This removes the dependency on selenium but requires a paid BS account that supports BS 

Finally we could opt for not testing this feature using Behat, accepeting the current limitation we have with the Goutte driver.

My preferred solution, as already made clear, would be to add a Selenium 2 driver with headless Google Chrome. This is a nice starting point for later adding BS support (also for testing WAYF & Consent features in greater detail and against multiple browsers)

See: 
* https://medium.com/caendra-tech/acceptance-tests-with-behat-and-chrome-headless-2824aaa31fd7
* https://www.browserstack.com/automate (Michiel Kodde - Mar 21, 2019)

**Comment by phavekes** [comment]
I\'ve started the implementation of supporting multiple authentication procedures. And I\'m running into a bit of a dilemma. The IdPController (Symfony) receives the AuthNRequest via HTTP Post or Redirect binding. So far so good.

Let me describe the workflow including the code responsible to give more insight into my dilemma.

For a solicited SSO request:
1. The AuthNRequest is verified in the IdPController (Symfony)
2. the SSO request is handled using the Corto Adapter (SingleSignOn), called from IdPcontroller (Corto)
3. The AuthenticationState is stored after receiving the AuthNRequest in the IdPController (Symfony)

At stage 2 we receive the SAMLRequest and further test if we can process it. Here we also can easily read the request id, which we need to track the authentication state for multiple in-flight authentications.

At stage 3, back in Symfony userland, we do no longer have access to the received SSO request. Unless we pull it from the session/inner state. Or again \'receive\' it from the HTTP request. Both options are not very elegant in my opinion.

My first thought was to migrate the functionality from the SingleSignOn Corto service to the Symfony application. This is still the best/future proof thing to do in my opinion. However, this would be a rather big change with possible side effects we need to consider and test.

Another, much more pragmatic solution would be to move step 3 from the Symfony IdPController to the SingleSignOn Corto service. This would defeat the need for refactoring, but break with our principle to stop putting new features into the legacy parts of EngineBlock.

My dilemma TL/DR: Moving newer symfony EB code to Corto seems counter productive, but prevents a potentially very big migration of Corto code into Symfony.

@bstrooband, @thijskh, @bartgeesink, and @pmeulen your feedback is most welcome! (Michiel Kodde - Mar 25, 2019)

**Comment by phavekes** [comment]
Thijs, Bas, and Michiel discussed the above just now on the telephone. A small summary of what we discussed and decided upon.

* Coupling to other Corto code is limited in the Corto services we targetted in the recent changes. This makes them a candidate for moving them to the \'Symfony side\'.
* For now, we will only move code if the impact to the rest of the EB code is low. We don\'t want to contaminate our sprint with unrelated refactor work.
* As a proof of concept, the ProcessConsent service will be moved to Symfony. Where we try to minimalize the changes to the actual code. Aside from implementing DI, and getting rid of super global use. Basically fixing the code to a bare minimum to allow PHPCS and PHPMD to pass. (Michiel Kodde - Mar 26, 2019)
